from flask import Flask, request, jsonify, render_template
from apscheduler.schedulers.background import BackgroundScheduler
from tweepy import Client
from flask_cors import CORS
import random, requests
import uuid
from dateutil.parser import parse
import os

app = Flask(__name__)
CORS(app)


# ËØªÂèñÈÖçÁΩÆÊñá‰ª∂
def read_config(file_path):
    config = {}
    with open(file_path, 'r', encoding='utf-8') as file:
        for line in file:
            if '=' in line:
                key, value = line.strip().split('=', 1)
                config[key] = value
    return config


# ËØªÂèñÈÖçÁΩÆ
config = read_config('config.txt')

# Twitter API ÂÆ¢Êà∑Á´Ø
twitter_client = Client(
    consumer_key=config['TWITTER_API_KEY'],
    consumer_secret=config['TWITTER_API_SECRET_KEY'],
    access_token=config['TWITTER_ACCESS_TOKEN'],
    access_token_secret=config['TWITTER_ACCESS_TOKEN_SECRET']
)


def translate_text(text):
    url = "https://api.siliconflow.cn/v1/chat/completions"
    payload = {
        "model": "Qwen/Qwen2.5-7B-Instruct",
        "messages": [
            {
                "role": "assistant",
                "content": "You are a skilled text processing expert with expertise in the corresponding field. Your function is to accurately process text according to instructions. Directly output the results, do not add any explanations or comments in the result text."
            },
            {
                "role": "user",
                "content": f"Perform academic polishing on the following source text, supplement and polish missing or unsmooth sentences, and directly output the processing results without any additional text. Output text: {text} Output text:"
            }
        ]
    }
    headers = {
        "Authorization": "Bearer sk-vgmfuoznkztusscvdkhffpophaahgrpjselkalfnfrvvmrkp",
        "Content-Type": "application/json"
    }
    response = requests.post(url, json=payload, headers=headers)
    if response.status_code == 200:
        return response.json()["choices"][0]["message"]["content"]
    else:
        return None


# ÂÆöÊó∂‰ªªÂä°Ë∞ÉÂ∫¶Âô®
scheduler = BackgroundScheduler(timezone="Asia/Shanghai")
scheduler.start()
# ÂàùÂßãÂåñË∞ÉÂ∫¶Âô®Âπ∂ËÆæÁΩÆÊó∂Âå∫‰∏∫‰∏≠ÂõΩÊó∂Âå∫ÔºàAsia/ShanghaiÔºâ

# ËØªÂèñÊñáÊú¨Êñá‰ª∂Âπ∂ÊåâÁ©∫Ë°åÂàÜÂâ≤ÊàêÊÆµËêΩ

paragraphs = [
    'By learning semantic relationships across multiple genes, Evo enables a\ngenomic ‚Äúautocomplete‚Äù in which a DNA prompt encoding a desired function instructs the model to generate\nnovelDNAsequencesthatcanbeminedforsimilarfunctions. Wetermthisprocess‚Äúsemanticmining,‚Äùwhich,\nunlike traditional genome mining, can access a sequence landscape unconstrained by discovered evolution-\nary innovation. We validate this approach by experimentally testing the activity of generated anti-CRISPR\nproteins and toxin-antitoxin systems, including de novo genes with no significant homology to any natural\nprotein.',
    'Strikingly, in-context protein design with Evo achieves potent activity and high experimental suc-\ncess rates even in the absence of structural hypotheses, known evolutionary conservation, or task-specific\nfine-tuning. We then use Evo to autocomplete millions of prompts to produce SynGenome, a first-of-its-kind\ndatabase containing over 120 billion base pairs of AI-generated genomic sequences that enables semantic\nmining across many possible functions. The semantic mining paradigm enables functional exploration that\nventures beyond the observed evolutionary universe.',
    '1. Introduction\nWhile generative AI promises to accelerate the design of functional biological systems, precisely articulating\n‚Äúfunction‚Äùtoagenerativemodelisachallengingandoftenunderspecifiedtask. Innaturallanguage,thenotion\nofdistributionalsemanticshypothesizesthatthemeaningofwordscanberepresentedbywordco-occurrence,\ni.e., that ‚Äúyou shall know a word by the company it keeps‚Äù (Firth, 1957;Harris,1954) (Figure1A ).',
    'In biology,\nan emerging distributional hypothesis defines the function of a gene through its interactions with other genes,\ni.e., you shall know a gene by the company it keeps (Kwon et al., 2024). In prokaryotic genomes, interacting genes are often positioned directly next to each other on the genome\nsequence in gene clusters or operons (Jacob and Monod, 1961;Overbeek et al., 1999). Researchers have\nexploited this property, in a process referred to as genome mining via guilt-by-association, to characterize\nunknown genes that neighbor functionally characterized genes (Aravind, 2000;Galperin and Koonin, 2000),\nresulting in the discovery of new molecular mechanisms (Lee et al., 2010;Pers et al., 2015) and important\nbiotechnological tools for genome editing, natural product synthesis, and RNA modification (Medema et al.,\n2011;Shmakov et al., 2015;Doron et al., 2018;Gao et al., 2020;Blin et al., 2021).',
    'At its core, guilt-by-\nassociation leverages the distributional hypothesis of gene function to perform function-guided discovery. A capable generative model of genomic sequences could learn a similar, distributional notion of function to\nperform function-guided design. Progress in long-context machine learning has enabled advanced generative\nmodels of genomic sequences at the multi-kilobase and multi-gene scale (Nguyen et al., 2024).',
    'It is made available under apreprint (which was not certified by peer review) is the author/funder, who has granted bioRxiv a license to display the preprint in The copyright holder for this this version posted December 18, 2024. ; https://doi.org/10.1101/2024.12.17.628962doi: bioRxiv preprint are trained to predict the next base pair in a sequence, enabling them to ‚Äúautocomplete‚Äù a partial genomic\nsequence by generating response sequences to a DNA sequence prompt (Figure 1B). Given the success of guilt-by-association, we reasoned that prompt engineering of a generative genomic\nmodel with a sequence of known function could instruct the model to sample novel, functionally related\ngenes in its response. We term this approach ‚Äúsemantic mining,‚Äù a new paradigm that harnesses multi-gene\nrelationships to generate novel DNA sequences enriched for targeted biological functions.',
    'Notably, unlike\ntraditional genome mining, semantic mining is unconstrained by known evolutionary innovation, potentially\nallowing us to explore entirely new regions of functional sequence space. Here, we show that Evo, a 7-billion parameter genomic language model, learns a distributional semantics\nover genes that enables function-guided design of proteins with high novelty. We first demonstrate that Evo\nenablesin-contextgenomicdesign,enablingthesuccessfulcompletionofpartialsequencesofhighlyconserved\ngenes and operons.',
    'We find that a new version of the previously described Evo model (Nguyen et al., 2024)\nthatwepretrainedon50%moredata,whichwerefertoasEvo1.5,obtainsthebestperformanceongene-level\nand operon-level completion tasks. We then apply semantic mining with Evo to design proteins with both high novelty and specified func-\ntional activity. We first focus on the generation of toxin-antitoxin systems, which requires designing two genes\nthat encode a protein-protein interaction.',
    'Using a sequential design strategy, we first generate novel toxins\nby prompting with genomic context around known toxin-antitoxin systems. We then use these Evo-generated\ntoxins as prompts to design their conjugate antitoxin pairs, yielding functional toxin-antitoxin systems con-\ntaining proteins with remote homology ( <30% sequence identity) to known proteins. We observed that 5 out\nof 10 designed proteins show antitoxin activity, representing a high experimental success rate of 50%.',
    'To assess Evo‚Äôs generalization capability beyond the design of known evolutionary homologs, we then\nperformed semantic mining of systems with high evolutionary novelty. In particular, we prompt-engineered\nEvo to generate anti-CRISPR proteins that, in nature, have high diversity in both sequence and mechanism\nand may be produced by de novo gene birth (Eitzinger et al., 2020;Huang et al., 2021;Dong et al., 2022). Strikingly, Evo generates anti-CRISPR proteins that effectively inhibit SpCas9-mediated DNA cleavage despite\npossessing no clear sequence homology to known proteins and low-confidence structure predictions.',
    'Evo also\ndesigns variants of known anti-CRISPR proteins with stronger inhibitory activity compared to potent natural\nanti-CRISPRs, as well as an inhibitor with sequence similarity to a protein family not previously associated\nwith CRISPR inhibition. For this task, we also observed a robust experimental success rate of 17%, with\n14 out of 84 designs demonstrating anti-CRISPR activity. Together, these results indicate that Evo may be\nable to generalize beyond the distribution of known natural protein sequences while retaining higher-level,\nuser-defined biological functions.',
    'Giventheexperimentalsuccessofsemanticminingingeneratingnovelfunctionalproteins,wethenuseEvo\ntogenerateSynGenome,thefirstAI-generatedgenomicsdatabase(https://evodesign.org/syngenome/),\ncontaining over 120 billion base pairs of synthetic DNA sequences derived from prompts spanning 1,705,667\nUniProt entries, 37,108 species, 72,492 protein domains, and 8,914 functional terms. The diversity of coding\nsequencesfoundinSynGenomerecapitulatesthenaturaldistributionofproteinfamilies,whilealsocontaining\nmany generations that go beyond the sequence landscape of the DNA prompts. We make SynGenome and Evo\n1.5 openly available to the scientific community to facilitate semantic mining across diverse functions.',
    'In total, Evo can successfully generate proteins with desired functions, often doing so by accessing a highly\nnovel space of sequences that are unlike any known proteins. Semantic mining, with its generalizability and\nrobust success rates, represents a promising new avenue for function-guided generative design. More broadly,\nwe anticipate this work to mark a turning point in biological discovery in which mining of genetic variation\nbecomes substantially augmented by AI-generated sequences that transcend known natural evolutionary con-\nstraints.',
    '2. CC-BY 4.0 International license perpetuity. It is made available under apreprint (which was not certified by peer review) is the author/funder, who has granted bioRxiv a license to display the preprint in The copyright holder for this this version posted December 18, 2024. ; https://doi.org/10.1101/2024.12.17.628962doi: bioRxiv preprint 2.',
    'Results\n2.1. Evo enables in-context genomic design\nToeffectivelyachievesemanticminingforfunction-guideddesign, amodelmustfirstunderstandnotjustindi-\nvidualgenesequences, buthowgenesrelatetoandinteractwitheachotherwithintheirbroadergenomiccon-\ntext. Evo, a 7-billion-parameter genomic language model trained on prokaryotic sequences, is well-positioned\nto address this challenge through its ability to process long genomic sequences at single-nucleotide resolution\n(Nguyen et al., 2024).',
    'Similar to how words in language derive meaning from their context (Figure 1A ), DNA sequences take\non functional significance when considered in the context of a gene, an operon, a pathway, or an entire\norganism. Evo‚Äôs efficient long-context architecture can learn how sequence patterns at the level of individual\nnucleotides relate to a genomic context containing multiple kilobases. Because functionally related sequences\ncluster together on prokaryotic genomes, supplying an appropriate functional context could condition Evo to\ngenerate new genetic sequences with a desired function (Figure 1B).',
    'As an initial experiment, we assessed Evo‚Äôs ability to leverage genomic context by performing an ‚Äúauto-\ncomplete‚Äùtaskinwhichwepromptedthemodelwithpartialsequencesofhighlyconservedprokaryoticgenes. We tested three diverse and functionally important genes from bacteria and archaea: RNA polymerase sigma\nfactor rpoSfrom E. coli, DNA gyrase subunit A gyrAfrom S. typhimurium , and cell division protein ftsZ1from\nH. volcanii (Dai and Lutkenhaus, 1991;Chiang and Schellhorn, 2010;Sada et al., 2022) (Figure 1C). We also\ntested three versions of the Evo model: Evo 1 8K, a model pretrained at 8,192 context length; Evo 1 131K, a\nmodel made by extending Evo 1 8K to 131,072 context length; and a model newly introduced in this study,\nwhich we call Evo 1.5, which extends the pretraining of Evo 1 8K (initially trained on 300 billion tokens) to\n450 billion tokens, representing a 50% increase in training data (Methods).',
    'For each gene, we prompted\neach Evo model version with varying amounts (30%, 50%, and 80%) of the input sequence and evaluated\neach model‚Äôs ability to complete the remainder of the gene. Evo 1.5 consistently demonstrated the highest recovery of the native protein sequence, particularly at\nlower prompt lengths. For instance, with just 30% of the input sequence, Evo 1.5 achieved approximately\n85% amino acid sequence recovery for rpoS, compared to 65% for Evo 1 131K.',
    'This performance advantage\nwas maintained across all tested genes and prompt lengths, with Evo 1.5 achieving near-perfect sequence\nrecovery at 80% input for all targets. These experiments are consistent with previous findings that longer\npretraining can improve learning of long-range interactions and high-level concepts in sequence models of\nnatural language and proteins (Lin et al., 2023;Kaplan et al., 2020;Brandes et al., 2022). Moving forward,\nwe therefore selected Evo 1.5 for further investigation and all results attributed to Evo in this study were\nproduced by the Evo 1.5 model.',
    'We further tested Evo‚Äôs understanding of genomic context at a multi-gene scale by evaluating its ability\nto predict gene sequences based on operonic neighbors (Figures 1D andS1). We prompted the model with\nsequences of genes either upstream or downstream of target genes in the well-characterized trpandmod-\nABCoperons (Maupin-Furlow et al., 1995;Merino et al., 2008), leveraging DNA complementarity to generate\nsequences bidirectionally. Evo demonstrated robust predictive performance across all tested configurations,\nachieving over 80% protein sequence recovery for all target genes.',
    'Further, the model exhibited adaptabil-\nity to genomic orientation, successfully generating downstream gene sequences when prompted with reverse\ncomplement sequences of upstream genes, and vice versa, while maintaining appropriate directionality across\ntheoperon. Notably, evenwithincompletesequencerecovery, thepredictedproteinstructuresofEvo‚Äôsgenera-\ntionscloselymatchedthenativeprotein‚Äôsconformations,suggestingpreservationoffunctionallycriticalamino\nacid residues and structural motifs. These results indicate that Evo not only learns the primary sequence of\ngenes but also captures the broader genomic organization of bacterial operons.',
    'To assess whether Evo‚Äôs generations go beyond trivial memorization of training sequences, we analyzed\nthe per-position entropy of both amino acid and nucleotide sequences in the model‚Äôs generations (Figure 1E). Using the modABC operon generations as a test case, we prompted the model with the sequence encoding\nmodAfrom E. coliK-12 and analyzed the variability in the generated modBresponses. The amino acid-level\nentropy analysis revealed selective conservation, with generally lower entropy at key structural positions and\nhigher variability in less conserved regions.',
    'This pattern aligns with natural protein evolution, where func-\ntionallycriticalresiduesshowhighconservationwhilenon-essentialpositionspermitaminoacidsubstitutions. 3. CC-BY 4.0 International license perpetuity.',
    'It is made available under apreprint (which was not certified by peer review) is the author/funder, who has granted bioRxiv a license to display the preprint in The copyright holder for this this version posted December 18, 2024. ; https://doi.org/10.1101/2024.12.17.628962doi: bioRxiv preprint Figure 1|In-context genomic modeling and design with Evo. (A) In natural language, distributional se-\nmantics leverages the property that functionally equivalent but lexically distinct words often occupy similar\npositionsinatextwithcommonsetsofneighboringwords. (B)Insemanticmining,agenomiclanguagemodel\ntrained across multiple genes learns to map functionally related genes to similar semantic spaces, enabling\ngeneration of functionally related but sequence-diverse genes.',
    '(C)Sequence recovery assessment when using\na genomic language model to autocomplete three conserved prokaryotic genes shows consistent improvement\ninperformancefromEvo1131KandEvo18KtoEvo1.5, demonstratingenhancedabilitytoleveragegenomic\ncontext. AA, amino acid; Bar height, mean; error bars, standard error; ùëõ= 100. (D)Bidirectional completion\nof conserved E. coli trp operon gene sequences demonstrates high sequence identity and structural conserva-\ntionacrosstheoperon.',
    '(E)PositionalentropycomparisonbetweenalignednaturalandalignedEvo-generated\nModBsequencesatnucleotideandaminoacidlevelsshowsconservationofessentialaminoacidresidueswhile\nmaintaining high nucleotide diversity. ùëõ= 500. 4.',
    'CC-BY 4.0 International license perpetuity. It is made available under apreprint (which was not certified by peer review) is the author/funder, who has granted bioRxiv a license to display the preprint in The copyright holder for this this version posted December 18, 2024. ; https://doi.org/10.1101/2024.12.17.628962doi: bioRxiv preprint At the nucleotide level, we observed substantially higher entropy across all positions, with variation even in\nregions coding for conserved amino acids. Given that a single prompt was used to generate all the response\nsequences, these results suggest that Evo is not simply reproducing memorized sequences but is rather syn-\nthesizing information from across its training set to reflect biological constraints while generating diversity in\na manner reminiscent of natural evolution.',
    '2.2. Semantic mining of functional protein-protein interactions with remote homology to nature\nThe ability of Evo to understand genomic context consisting of highly conserved genes and operons led us\nto explore if we could apply semantic mining to less well-conserved biology: phage and bacterial defense\nsystems. These systems, which are shaped by the unrelenting evolutionary arms race between bacteria and\nphage(Murtazalievaetal., 2024),aresomeofthemostrapidlyevolvingsystemsinnature.',
    'Asaresult,defense\nsystems exhibit vast amounts of functional diversity while maintaining minimal sequence conservation across\nspecies (Beavogui et al., 2024;Tesson et al., 2024). Interestingly, defense systems of similar functionality also\nfrequently cluster into defense islands, enabling the discovery of novel systems through guilt-by-association\napproaches (Makarova et al., 2011;Koonin, 2018). To date, some of the most fruitful discoveries of genome\nmininghavebeenderivedfromdefensesystems, includingCRISPR-Cassystems, restrictionenzymes, andvari-\nousantimicrobialcompounds(Doronetal., 2018;Gaoetal., 2020;Pictonetal., 2021;Altae-Tranetal., 2023;\nCheng et al., 2024).',
    'The remarkable success in repurposing these naturally occurring systems for biotechnol-\nogyapplicationsunderscoresthevastpotentialofbacterialdefensemechanismsasasourceofmoleculartools. Given thenatural diversityand genomicco-localization ofthese systems, we soughtto determineif seman-\ntic mining could be used to design new defense systems. As an initial test case, we focus specifically on type II\ntoxin-antitoxin (T2TA) systems, some of which play a role in phage defense.',
    'Type II toxin-antitoxins consist\nof a stable toxin protein that can inhibit cell growth or induce cell death under conditions of stress, such as a\nphage infection, paired with an unstable antitoxin protein that binds and neutralizes the toxin‚Äôs activity under\nhomeostatic conditions ( Figures 2A,B). These systems often maintain conserved high-level genomic archi-\ntecture despite sequence divergence, with toxin and antitoxin genes typically arranged in adjacent positions\n(Chan et al., 2023;Guan et al., 2024). To generate diversified T2TAs using Evo 1.5, we developed a prompting strategy that leveraged the charac-\nteristicco-localizationofthesesystems( Figure2A ).Inadditiontopromptingwithsequencesencodingknown\nT2TAs,wecompiledgenomicsequencesdirectlyupstreamanddownstreamofknownT2TAsystems.',
    'Following\nsampling using the compiled prompts, we filtered generations for sequences encoding toxin-antitoxin protein\npairs that exhibited in silico predicted protein complex formation with AlphaFold 3 (Abramson et al., 2024)\nyet had limited sequence homology to known T2TA proteins. We first tested Evo-generated toxins using a growth inhibition assay, in which growth arrest indicated\nsuccessful toxin activity (Methods). From these experiments, we were able to identify a functional bacte-\nrial toxin, which we call EvoRelE1, that exhibited strong growth inhibition (approximately 70% reduction in\nrelative survival) while possessing 71% sequence identity to a known RelE toxin (Figures 2C,D).',
    'WesubsequentlypromptedEvo1.5withthesequenceofEvoRelE1, hypothesizingthatthemodelcoulduse\nthe genomic context of the toxin sequence to generate a conjugate antitoxin ( Figure 2A ). Using this strategy,\nwe identified a set of 10 antitoxin candidates exhibiting little to no sequence homology to natural protein\nsequences that were tested for their ability to block the toxin activity of EvoRelE1. Following co-expression with EvoRelE1, half of the Evo-generated antitoxin candidates were able to rescue\ncell growth (Figures 2D andS2).',
    'The most effective candidates, designated as EvoAT1 and EvoAT2, restored\ngrowth to 95-100% of normal cell survival, with candidates EvoAT3 and EvoAT4 demonstrating moderate\nrescue activity (70% and 90% relative survival, respectively). Sequence analysis of the successful antitoxins\nrevealed discrete regions of conservation (Figure 2E ), potentially highlighting key motifs required for toxin\nneutralization despite their overall sequence diversity. Furthermore, when tested against native RelE, MazF,\nandYoeBtoxinsystems,severaloftheEvo-generatedantitoxinswereabletorescuegrowth,withEvoAT2show-\ninginhibitoryactivityagainstallthreetoxinsandEvoAT4rescuinggrowthagainstnativeRelEandYoeBtoxins\n(Figure 2F).',
    'This observation suggests that the EvoATs may be able to inhibit multiple toxins despite shar-\ning limited sequence homology with their natural antitoxin counterparts (Figures 2F andS3). Interestingly,\nwhen co-folded using in silico structure prediction methods, several of the EvoATs had low-confidence pre-\n5. CC-BY 4.0 International license perpetuity.',
    'It is made available under apreprint (which was not certified by peer review) is the author/funder, who has granted bioRxiv a license to display the preprint in The copyright holder for this this version posted December 18, 2024. ; https://doi.org/10.1101/2024.12.17.628962doi: bioRxiv preprint Figure 2|Evo generates functional toxin-antitoxin protein-protein interactions with remote homology\nto nature. (A) To perform semantic mining of toxin-antitoxin systems, toxins (T) are first generated using\ntoxin genomic context as prompts and tested via growth inhibition assay. Successful toxins (EvoT) then serve\nas prompts to generate cognate antitoxins (AT), which are finally validated through a growth recovery assay.',
    '(B)Type II toxin-antitoxin systems function via antitoxin binding and neutralization of toxin activity to pre-\nvent cell death. (C)AlphaFold 3 structure prediction comparison between functional toxin EvoRelE1 and its\nclosest sequence match. pLDDT, predicted local distance difference test score; TM-score, template modeling\nscore.',
    '(D)Bar plot showing relative survival rates for different Evo-generated toxin-antitoxin combinations. EvoAT1-4 achieve high growth rescue against EvoRelE1. Bar height, mean; error bars, standard error; circles,\nbiological replicate values; ùëõ= 3.',
    'Statistical significance determined by one-sided Student‚Äôs t-test comparing\nto EvoRelE1+eGFP (*** ùëù< 0.0025) (E)Homology analysis of aligned amino acid sequences for EvoAT1-4,\nwith darker green indicating higher conservation. (F)Heat map showing relative survival rates of EvoAT1-4\ntested against native toxins. EvoAT2 and EvoAT4 appear to have inhibitory activity against multiple natural\ntoxins.',
    '(G)AlphaFold 3 structure predictions of EvoAT1-4 in complex with EvoRelE1 and structure compar-\nisons between Evo-generated antitoxins and their closest natural matches. EvoAT1-4 exhibit high confidence\npredicted structures despite limited sequence homology to their closest protein matches. (H)AlphaFold 3\npredicted aligned error plots for EvoAT1-4 in complex with EvoRelE1 showing high confidence in predicted\nstructures and interactions.',
    '(I)Heatmap showing number of structural and sequence matches between Evo-\ngenerated antitoxins and known antitoxin families. B, BLAST; F, FoldSeek; H, HHpred. 6.',
    'CC-BY 4.0 International license perpetuity. It is made available under apreprint (which was not certified by peer review) is the author/funder, who has granted bioRxiv a license to display the preprint in The copyright holder for this this version posted December 18, 2024. ; https://doi.org/10.1101/2024.12.17.628962doi: bioRxiv preprint dicted complex formation with the natural toxins that they inhibited (EvoAT4 + YoeB ipTM = 0.14, EvoAT2\n+ YoeB ipTM = 0.47, EvoAT2 + MazF ipTM = 0.34) ( Figure S4), highlighting the potential for semantic\nmining to identify novel molecular interactions that extend beyond the domain of state-of-the-art structure\nprediction models. EvoAT1throughEvoAT4allhavesequenceidentitieswithnaturalproteinsthatfallnearthe‚Äútwilightzone‚Äù\nof sequence similarity (20% to 36%), a range where sequence similarity alone is insufficient to reliably pre-\ndict shared structure or function (Rost, 1999).',
    'Both EvoAT1 and EvoAT2 only exhibit homology to proteins\nnot actively annotated as antitoxins, showing just 26% sequence identity to a DUF6290 family protein from\nActinomyces and 24% sequence identity to an uncharacterized Magnetococcus sp. YQC-5 hypothetical protein\nrespectively (Figure 2G). Given that EvoAT2‚Äôs closest match is an uncharacterized hypothetical protein that\nappears in a similar genomic context ( Figure S5), our results suggest this protein may function as part of a\ntoxin-antitoxin system.',
    'Similarly, the homology between EvoAT1 and the DUF6290 family reinforces a hy-\npothesized potential role for this domain of unknown function in antitoxin activity (Blum et al., 2024). These\nfindings suggest that Evo‚Äôs understanding of genomic context enables functional annotation of previously un-\ncharacterized proteins, as well as function-conditioned design that is unconstrained by existing annotations. The structural predictions for EvoAT1 and EvoAT2 revealed high-confidence folds (pLDDT scores of 0.85 and\n0.83, respectively) (Figure 2G), and when co-folded with EvoRelE1, both antitoxins exhibited minimal pre-\ndicted position error (Figure2H).',
    'These structural characteristics and high functional activity are particularly\nnoteworthy given the antitoxins‚Äô limited sequence homology to known antitoxin proteins (Figure2I). Overall,\nthese results further underscore the ability of Evo to generate functional proteins with remote homology to\nnatural proteins. SequenceandstructuralhomologysearchesusingBLAST,HHpred,andFoldseek(vanKempenetal., 2024;\nS√∂dingetal., 2005;Gableretal., 2020;Zimmermannetal., 2018;Sayersetal., 2022)revealedthattheEvoATs\nshow similarity to multiple independent antitoxin superfamilies, particularly ParD, MazE, and VapB (Figure\n2I).',
    'This finding, coupled with the activity of EvoAT2 and EvoAT4 against multiple toxins, is especially no-\ntable because their cognate toxins employ fundamentally different mechanisms of action. In natural systems,\nthese mechanistic differences have typically led to the evolution of distinct antitoxin structures specialized\nfor neutralizing each type of toxin (Fraikin et al., 2020;Qiu et al., 2022). This suggests that Evo may have\nidentifiedabroaderfunctionalcompatibilitybetweenantitoxinandtoxinfamiliesthanpreviouslyrecognized,\nhighlighting the potential of semantic mining to reveal new insights into protein-protein interaction patterns\nin prokaryotic systems.',
    '2.3. Semantic mining of functional de novogenes\nGiventheexperimentalsuccessofEvo-generatedfunctionalproteinswithremotehomologytonature, wenext\nexplored whether semantic mining could learn to generate even greater evolutionary novelty. Anti-CRISPRs\n(Acrs)areproteinsusedbyphagestoneutralizebacterialCRISPR-Casimmunesystems.',
    'Giventheircriticalrole\nin helping phages evade CRISPR-Cas mediated cleavage (Figure 3A ), Acrs represent one of the most striking\nexamples of rapid protein evolution in bacterial-phage arms races, with many Acrs appearing to be novel\ninnovationslackingdetectablehomologytootherproteinfamilies,includingotherAcrs(Davidsonetal., 2020;\nEitzinger et al., 2020;Huang et al., 2021). This remarkable diversity is reflected in their varied mechanisms\nof action, from direct CRISPR-Cas protein binding to DNA mimicry to transcriptional silencing (Shin et al.,\n2017;Camara-Wilpert et al., 2023;Choudhary et al., 2023;Trost et al., 2024), making them valuable tools for\nunderstanding protein evolution and developing synthetic control systems for CRISPR technologies (Pawluk\net al.,2018). Despite their sequence diversity, many Acr operons maintain a somewhat conserved operonic architecture,\nconsistingofmultipleAcrgenesappearingtogetheralongsideneighboringanti-CRISPRassociated(aca)genes\n(Stanley et al., 2019) (Figure 3B).',
    'This architectural conservation, combined with their frequent emergence\nasde novogenes,makesAcrsanidealtestcaseforassessingsemanticmining‚Äôsabilitytogeneralizewithrespect\nto sequence and structure while retaining a desired higher-level function. Using genomic sequences from known Cas9-targeting Acr genes and their associated operons as prompts,\nwe used Evo 1.5 to generate 3,160 sequences totaling 3.16 million base pairs. After filtering for size, com-\nplexity, and secondary structure, we identified 468 potential Acr open reading frames (ORFs).',
    'To handle the\nhigh sequence diversity of Acr proteins complicating traditional MSAand structural filtering methods, we next\n7. CC-BY 4.0 International license perpetuity. It is made available under apreprint (which was not certified by peer review) is the author/funder, who has granted bioRxiv a license to display the preprint in The copyright holder for this this version posted December 18, 2024. ; https://doi.org/10.1101/2024.12.17.628962doi: bioRxiv preprint Figure 3|Evo generates functional anti-CRISPR proteins with no homology to known proteins.',
    '(A) Type\nII anti-CRISPR systems involve an anti-CRISPR (acr ) gene that encodes an Acr protein that inhibits type II-A\nCas nuclease activity, often co-occurring with anti-CRISPR associated genes ( aca). (B)To perform semantic\nmining of Acrs, known genomic contexts of type II anti-CRISPRs, including potential acagenes, are used as\nprompts. (C)PaCRISPRclassificationdemonstratessignificantenrichmentofAcr-likesequencesingenerations\nfromAcrpromptscomparedtocontrols.',
    'Barheight,mean; errorbars,standarderror; circles,differentbatches\nof generations; ùëõ= 4. (D)Sequence identity matrix demonstrating high diversity among random sample of\nEvo-generatedAcrs. (E)Intheanti-CRISPRprotectionassay,functionalAcrproteinspreventSpCas9-mediated\ncleavageofakanamycinresistancegene, enablingcellsurvivalinkanamycin-supplementedmedia, whilenon-\nfunctional candidates allow SpCas9 targeting and cell death.',
    '(F)Bar plot showing relative survival rates for\nEvo-generated Acrs. EvoAcr1-5 achieve robust protection against SpCas9, with EvoAcr5 exceeding the activity\nof the AcrIIA2 positive control. Bar height, mean; error bars, standard error; circles, biological replicate\nvalues;ùëõ=3.',
    'Statisticalsignificanceisdeterminedbyaone-sidedStudent‚Äôst-testcomparedtorandomcontrol\n(***ùëù< 0.0025) and AcrIIA2 (ÀÜ ùëù< 0.0025). (G)T4 phage plaque assays validating anti-CRISPR activity. E.\ncolicells co-expressing SpCas9 targeted to T4 phage and candidate Acrs were embedded in soft agar, and\nserial dilutions of T4 phage were spotted onto the bacterial lawn.',
    'Plaque formation indicates successful anti-\nCRISPR protection. Experiment performed in triplicate with representative images shown. (H)AlphaFold\n3 structure predictions for EvoAcr1-2 showing low confidence predictions.',
    'pLDDT, predicted local distance\ndifference test score; TM-score, template modeling score. (I)AlphaFold 3 predicted structure of EvoAcr3\nand structure alignment between EvoAcr3 and its closest sequence match, revealing limited homology to a\nsigma-70 family protein. (J)AlphaFold 3 structure predictions comparing EvoAcr4-5 with their closest Acr\nsequence matches, showing moderate structural and sequence similarity.',
    '(K)Sequence similarity analysis of\nEvoAcr1-5calculatedbyBLASTsearchagainstBLAST-nrandMMseqs2searchagainstOpenGenome. EvoAcr1-\n2 demonstrate no significant homology to known proteins, EvoAcr3 shows matches only in the ‚Äútwilight zone‚Äù\nof sequence similarity (Rost, 1999), and EvoAcr4-5 show limited sequence matches to known Acrs. 8.',
    'CC-BY 4.0 International license perpetuity. It is made available under apreprint (which was not certified by peer review) is the author/funder, who has granted bioRxiv a license to display the preprint in The copyright holder for this this version posted December 18, 2024. ; https://doi.org/10.1101/2024.12.17.628962doi: bioRxiv preprint used PaCRISPR, a machine learning model trained to identify potential Acr proteins, to evaluate our gener-\nated candidates (Wang et al., 2020). Consistent with our prompts providing successful functional condition-\ning, generations derived from Acr-containing genomic contexts were substantially more likely to be classified\nas potential Acrs by PaCRISPR compared to negative control sequences (Figure 3C).',
    'Approximately 55% of\nminimally filtered proteins generated from Acr-associated prompts were classified as likely Acrs, compared\nto less than 5% of proteins generated from either random prompts or random genomic sequences. Further,\nthe distribution of sequence identities among the predicted Acrs showed a wide range of novelty, with most\ncandidates showing low similarity to each other (Figure 3D) (median pairwise sequence identity = 23%). This enrichment for diverse Acr-like sequences demonstrates that semantic mining can effectively bias gener-\nations toward a desired function by leveraging genomic context alone, even in the absence of clear sequence\nor structural signatures.',
    'TotesttheAcrcandidates‚ÄôprotectionabilityagainstSpCas9,weco-transformed E. coliwithplasmidsencod-\ning our candidate Acrs and a CRISPR-targeted kanamycin resistance gene, where functional Acr protection\nwould enable bacterial growth in kanamycin-supplemented media through inhibition of CRISPR-mediated\ncleavage (Forsberg et al., 2019) (Figure 3E). From this initial experiment, we found that 17% of tested se-\nquences exhibited some degree of anti-CRISPR activity (Figure S6), a high success rate when considering the\nlack of structure or sequence homology-guided design. From this pool, we further identified five promising\nproteins (EvoAcr1-5) that demonstrated strong protection against SpCas9-mediated targeting in both liquid\nculture survival assays (Figure 3F) and phage infection experiments (Figures 3G andS7).',
    'Detailed bioinformatic analysis of these five Acrs revealed a high level of diversity in their sequence ori-\ngins. EvoAcr4 and EvoAcr5 share moderate sequence similarity to known Acrs (61% and 48% to AcrIIA2\nand AcrIIA4, respectively) and demonstrated robust protection against SpCas9, with EvoAcr5 showing activity\nexceeding that of the positive control AcrIIA2 in liquid culture assays (relative survival rates of 1.01 to 0.87\nrespectively) (Figures 3F,H). EvoAcr3 presented an intriguing case: while sharing only ‚Äútwilight zone‚Äù se-\nquence identity (Rost, 1999) and limited structural alignment (sequence ID = 25%, TM-score = 0.29) with\na sigma-70 family RNA polymerase sigma factor, it maintained strong anti-CRISPR activity (relative survival\nof 0.89) (Figures 3F,I).',
    'This suggests a potential mode of CRISPR inhibition that is not described by existing\nfunctional databases. Most notably, EvoAcr1 and EvoAcr2 represented proteins that eluded both sequence and structural homol-\nogy characterization, showing no significant identity to proteins in OpenGenome (the Evo training data) or to\nknown proteins in the nr database by BLAST (Figures3J,K ). The AlphaFold 3-predicted structures of EvoAcr1\nand EvoAcr2 also have low confidence pLDDT scores of 0.46 and 0.36, respectively (Figures 3H andS8).',
    'De-\nspite this lack of sequence or structural similarity to known proteins, they demonstrated robust protection in\nboth liquid culture and phage infection assays, with relative survival rates of 0.82 and 0.74, respectively. This\nexperimental validation of novel, functional Acrs supports the ability of semantic mining to access entirely\nunexplored regions of sequence space while maintaining specific desired functions. Together with EvoAcr3-5,\ntheseresultsdemonstratethatgenomiccontextcanguidethegenerationofdiverseanti-CRISPRproteins,from\nsequence variants of known architectures to entirely new protein sequences.',
    '2.4. SynGenome: A first-of-its-kind synthetic genomic database\nFollowing our experimental validation that semantic mining can generate novel, functional proteins from ge-\nnomic context alone, we reasoned that this approach could be systematically applied to generate functional\ngenes across prokaryotic biology. To this end, we developed SynGenome, a comprehensive database of syn-\nthetic DNA sequences generated using Evo.',
    'In line with the function-guided design principles underlying se-\nmantic mining, we prompted the model with sequences derived from over 1.5 million prokaryotic and phage\nproteinstogeneratealargesetofdiversesequencesthatcouldbeenrichedfornovelgenesfunctionallyrelated\nto the prompt. To construct SynGenome, we leveraged the UniProt database to identify protein-coding genes and their\nadjacent sequences from prokaryotic organisms and bacteriophages. For each coding sequence, we generated\nsix distinct prompts: the upstream region, coding sequence (CDS), downstream region, and their respective\nreverse complements (Figure 4A ).',
    'Using the Evo 1.5 model with standard autoregressive decoding, we gen-\nerated multiple synthetic sequences for each prompt, resulting in a database containing over 120 billion DNA\nbase pairs (Methods). 9. CC-BY 4.0 International license perpetuity.',
    'It is made available under apreprint (which was not certified by peer review) is the author/funder, who has granted bioRxiv a license to display the preprint in The copyright holder for this this version posted December 18, 2024. ; https://doi.org/10.1101/2024.12.17.628962doi: bioRxiv preprint Figure 4|120 billion base pairs of AI-generated genomic sequences with SynGenome. (A) To construct\nSynGenome, we created prompts from known-protein coding genes, generated synthetic DNA sequences us-\ning Evo, and bioinformatically characterized the generated sequences. (B)Number of prompts and gener-\nated sequences in SynGenome, and their associated features.',
    'GO, gene ontology. (C)Codon usage bias of\nProdigal-predicted ORFs in prompt sequences and generated sequences, demonstrating preservation of codon\npreferences. (D)UMAP of jointly projected Evo embeddings of SynGenome and prompt sequences.',
    'Genera-\ntion embeddings are colored by sequence identity to training data and prompt embeddings are shown in light\ngreen. (E)BarplotshowingrelativeproportionsofgenerationandpromptsequencesinmostpopulousLeiden\nclusters, with percentages indicating fraction of generated sequences per cluster. (F)Prodigal-predicted ORF\nlengths in for 5000 nt representative samples of OpenGenome and SynGenome, with ORFs showing similar\nlength distributions.',
    'nt, nucleotide. (G)Histogram showing number of Pfam protein family occurrences in\nSynGenome and OpenGenome, with both following a similar long-tailed distribution of family abundance. (H)Scatterplot showing correlation between protein family frequencies in SynGenome versus OpenGenome,\ndemonstrating preservation of natural protein family abundance patterns.',
    '10. CC-BY 4.0 International license perpetuity. It is made available under apreprint (which was not certified by peer review) is the author/funder, who has granted bioRxiv a license to display the preprint in The copyright holder for this this version posted December 18, 2024. ; https://doi.org/10.1101/2024.12.17.628962doi: bioRxiv preprint To facilitate functional exploration of the database, we organized the synthetic sequences according to\nthe Gene Ontology (GO) terms and InterPro domain annotations associated with their corresponding prompts\n(Figure 4B), as the generated sequences are likely to be enriched for functionally related elements.',
    'As gener-\nating sequences with language models can often be prone to repetitive generations, we implemented minimal\nfiltering to remove highly repetitive low complexity sequences, for example, long stretches of sequence con-\ntaining only a single base pair (Methods). To characterize the generations in SynGenome, we first examined codon usage patterns between the gen-\nerated sequences and the prompts. This analysis revealed that generated sequences closely mirrored their\nprompts, with the synthetic sequences maintaining similar codon preferences to natural coding sequences\n(Figure 4C).',
    'We further examined the relationship between prompts and their corresponding generated se-\nquences in Evo embedding space (Figure 4D) by performing Leiden clustering and analyzing the relative\nproportions of each sequence type within clusters (Figures 4E andS9). Most Leiden clusters contain a mix-\nture of generated sequences and natural prompt sequences, indicating that many generations occupy similar\nregions of the language model embedding space. However, we observed that 54 clusters representing 19%\nof the generated sequences are almost completely composed of generated sequences, potentially indicating\nregions of synthetic sequences that extend beyond the semantic space represented by natural sequence em-\nbeddings.',
    'When compared to sequences from OpenGenome, we found that SynGenome-generated ORFs closely mir-\nrored the natural prokaryotic ORF length distribution predicted by Prodigal (Hyatt et al., 2010) (Figure 4F). At the protein level, SynGenome matched both the global distribution of Pfam domain frequencies found in\nnaturalproteins(Figure4G),withspecificPfamfamiliesalsoshowingsimilaroccurrencefrequenciesbetween\nsynthetic and natural sequences (Mistry et al., 2021;Blum et al., 2024) (Pearson correlation coefficient, r=\n0.78, (Figure4H). These analyses demonstrate that SynGenome recapitulates the architecture and functional\ndiversity of natural prokaryotic sequences.',
    'Collectively, these results highlight the potential for SynGenome to become a valuable tool for exploring\nand expanding protein function through semantic relationships. By mining SynGenome, researchers could\ndiscover functionally related proteins that extend beyond natural sequence space, gain insights into potential\nfunctions of uncharacterized proteins based on genetic co-occurrence patterns, and create diverse screening\nlibraries for engineering desired functions (Figure S10). SynGenome is freely available as a public resource at https://evodesign.org/syngenome/, where\nresearchers can search the database using protein names, UniProt IDs, InterPro domains, species names, or\nGO terms of interest.',
    'This allows for rapid identification of relevant synthetic sequences, enabling researchers\nto quickly find and explore generations that align with their research interests. We anticipate that SynGenome\ncan serve as a practical tool that facilitates gene discovery and engineering with semantic mining for the\nbroader scientific community. 3.',
    'Discussion\nAdvanced genomic sequence models, trained on hundreds of billions of DNA base pairs across the diversity of\nprokaryotic life, can enable unprecedented capabilities. We demonstrate that Evo enables controllable design\nof desired functions when prompted with sequences that are meant to elicit those functions, with high exper-\nimental success rates of 17-50% when testing only tens of variants in the laboratory. Many of the designed\nproteins also have no significant homology to proteins of similar function or, in some cases, to any known\nprotein.',
    'These results blur the line between ‚Äúde novo‚Äù protein design (Huang et al., 2016;Watson et al., 2023;\nKortemme, 2024) and diversification based on evolutionary models (Madani et al., 2023;Hayes et al., 2024;\nHie et al., 2024), providing an ‚Äúexistence proof‚Äù that sequence models without any structural conditioning\ncan meaningfully generalize beyond natural sequence space. Semantic mining represents a fundamentally new paradigm for protein design that is highly complemen-\ntary to existing approaches. First, unlike design methods based on task-specific fine-tuning (Madani et al.,\n2023;Jiang et al., 2024;Nguyen et al., 2024), semantic mining requires no additional training or supervision\nthat could bias generations toward the distribution of known examples.',
    'Second, in contrast with approaches\nthat propose specifying function through natural language descriptions found in existing knowledgebases\n(Praljak et al., 2024), semantic mining accesses the vast functional diversity embedded within genomic se-\n11. CC-BY 4.0 International license perpetuity. It is made available under apreprint (which was not certified by peer review) is the author/funder, who has granted bioRxiv a license to display the preprint in The copyright holder for this this version posted December 18, 2024. ; https://doi.org/10.1101/2024.12.17.628962doi: bioRxiv preprint quences.',
    'This enables functional design that can leverage biological processes that are not yet documented\nin the scientific literature. For example, we generate antitoxins that suggest broader functional compatibil-\nity between diverse toxin and antitoxins systems than previously recognized (Figure 2F) and a protein with\nanti-CRISPR activity that maps to a protein family with a different putative function (Figure 3I). Third, by\nleveraginggenomiccontextasfunctionalconditioning, semanticminingdoesnotrequireexistingstructuralor\nmechanistic hypotheses; indeed, protein design pipelines that filter out low-confidence structure predictions\n(Verkuil et al., 2022;Pacesa et al., 2024) would have removed most of our high-activity designs.',
    'Genomic\nconditioning is also useful when specifying functions such as anti-CRISPR activity that could be accomplished\nby many structures and mechanisms (Choudhary et al., 2023). Semantic mining of genomic language models\ntherefore represents a powerful and orthogonal approach to current biological design strategies. Traditionalgenomeminingviaguilt-by-association, whichmotivatesmanyoftheideasinthisstudy, iscon-\nstrained to discovered evolutionary diversity that was generated over billions of years.',
    'In contrast, semantic\nmining enables the generation of a near limitless supply of sequence diversity for a biological system of inter-\nest. To facilitate broader accessibility of this new source of evolutionary material, we report SynGenome, a\ndatabase of over 120 billion base pairs of AI-generated genomic sequences, which we make publicly available. This resource enables researchers, especially those without the computing resources required to conduct large\nscale sampling from a generative model, to mine for sequences related to their function of interest.',
    'This data\ncould potentially contain new molecular tools and provide insights into protein function and evolution. While semantic mining represents a new level of sequence novelty and functional improvement for gen-\nerative genomics, several fundamental limitations and challenges remain. Autoregressive generation is prone\nto sampling repetitive sequences or to hallucinating realistic but non-functional designs.',
    'Semantic mining\ntherefore requires both in silicofiltering and experimental testing to validate downstream functions. Semantic\nmining is also theoretically limited to functions that are found in nature, particularly those that are encoded\nin prokaryotic organisms. However, we note that only a small fraction of this functional diversity has been\ncharacterized by the biological literature and that mining of prokaryotic functional diversity has directly led\nto powerful biotechnologies such as polymerase chain reaction (PCR), optogenetics, genome editing, and ge-\nnetic recombination (Chien et al., 1976;Nagel et al., 2003;Boyden et al., 2005;Govorunova et al., 2017;\nAltae-Tran et al., 2023;Durrant et al., 2024).',
    'Looking forward, increased scaling of pretrained models and availability of training data from rapidly\ngrowing sequencing resources will likely reinforce the capabilities of semantic mining. We also anticipate\ncombining the rich information learned by pretrained models with more advanced inference-time strategies\nwill improve generation quality. Analogous to generative models of images that exhibit improved controllabil-\nity when given prompts that are themselves generated by a natural language model (Esfandiarpoor and Bach,\n2024;Ma√±as et al., 2024), synthetic prompts could be used to improve the generation quality of genomic\nlanguage models as well; for example, EvoAT1-4 were all generated in the context of the previously designed\nEvoRelE1.',
    'Further, given the utility of chain-of-thought prompting in natural language (Wei et al., 2023),\ngenomic language models that sequentially generate multicomponent systems could accelerate the develop-\nment of synthetic biological circuits, molecular machines, metabolic pathways, or even complete genomes. Lastly, techniques for mechanistic interpretability of large language models (Bricken et al., 2023;Temple-\nton et al., 2024) or post-training that leverages high-throughput experimental data collection (Ouyang et al.,\n2022;Rafailov et al., 2024;Rai et al., 2024) could improve the quality and diversity of generated functions. By leveraging semantic mining, exploration of synthetic genomic space may reveal biological discoveries that\ncomplement and extend beyond those discovered in natural organisms.',
    'Code availability\nCodeforgenerationwithEvobasedonauser-definedpromptisavailableat https://github.com/evo-design/\nevo/. Scripts for semantic mining of toxin-antitoxin systems and anti-CRISPRs are available https://\ngithub.com/evo-design/evo/tree/main/semantic_mining. The Evo 1.5 model is available under\nan open source license at Hugging Face: https://huggingface.co/evo-design/evo-1.5-8k-base.',
    '5. Data availability\nSynGenomeisexplorableandsearchableat https://evodesign.org/syngenome/. Rawdatahasbeende-\npositedtoHuggingFacedatasetsat https://huggingface.co/datasets/evo-design/syngenome-uniprot.',
    '6. Materials availability\nDNA, RNA, and protein sequences used during our validation experiments are available in Data S1. All newly\ncreated materials are available upon reasonable request to the corresponding authors.',
    '7. Acknowledgements\nWe thank Kevin Forsberg, Regan Russell, and Samantha Sakells for supplying plasmid backbones and for\nhelpfuldiscussionsonanti-CRISPRexperiments. WethankMatthewDurrant, PatrickHsu, JuliaKazaks, David\nLi,TalalWidatalla,BrandonAmeglio,SergeyOvchinnikov,AprilPawluk,BrianPlosky,andChiaraRicci-Tamfor\nhelpful discussions and assistance with manuscript preparation.',
    'A.T.M\ncompiled prompts for the toxin-antitoxin sampling. A.T.M developed code for and performed the analysis and\nfiltering of the toxin-antitoxin candidates. A.T.M experimentally tested the toxin-antitoxin generations.\nA.T.M\ncompiled the prompts for SynGenome. A.T.M and B.L.H. developed the code for and performed the sampling\nfor SynGenome.\nB.L.H. developed the SynGenome website. A.T.M.',
    'S.H.K., and B.L.H. wrote the first draft of\nthe manuscript. All authors wrote the final draft of the manuscript.\n9. Competing Interests\nB.L.H. acknowledges outside interest in Prox Biosciences as a scientific co-founder.',
    'A.T.M, S.H.K., and B.L.H. are named on a provisional patent application applied for by Stanford University and Arc Institute related to\nthis manuscript. All other authors declare no competing interests.\nAll other authors declare no competing\ninterests. 13. CC-BY 4.0 International license perpetuity.',
    'It is made available under apreprint (which was not certified by peer review) is the author/funder, who has granted bioRxiv a license to display the preprint in The copyright holder for this this version posted December 18, 2024. ; https://doi.org/10.1101/2024.12.17.628962doi: bioRxiv preprint 10. Methods\nEvo 1.5 Pretraining\nWe extended the pretraining of the Evo 1 model that was trained at a sequence context of 8,192 tokens with\nan initial learning rate of 0.003 after a warmup of 1,200 iterations; a cosine decay schedule with a maximum\ndecay iterations of 120,000 and a minimum learning rate of 0.0003; a global batch size of 4,194,304 tokens;\nand 75,000 iterations (processing a total of 315 billion tokens). Other details on hyperparameters related\nto the model architecture and optimizer can be found in Nguyen et al.',
    '(Nguyen et al., 2024). We resumed\nthe model‚Äôs pretraining, including all model states, optimizer states, data loading schedule, and learning rate\nschedule, from 75,000 iterations to 112,000 iterations (processing a total of 470 billion tokens). We refer to\nthe model trained up to 112,000 iterations as Evo 1.5.',
    'Autoregressive sampling\nTo sample from Evo, we used a standard low-temperature autoregressive sampling algorithm (Chang and\nBergen, 2023). We used the sampling code at https://github.com/evo-design/evo/ based on the ref-\nerence implementation at https://github.com/togethercomputer/stripedhyena that leverages kv-\ncaching of Transformer layers and the recurrent formulation of hyena layers (Massaroli et al., 2023; Poli et al.,\n2023) to achieve efficient, low-memory autoregressive generation. Parameter optimization was performed\nacross various temperatures (0.1-1.5, increments of 0.1), top-k values (1-4), and top-p values (0.1-1.0, incre-\nments of 0.1) using the Evo 1.5 model.',
    'For each parameter combination, one hundred 1,000-nt sequences\nwere generated from a test set of 5 gene prompts encoding 50% of a highly conserved protein. Following\nidentification of ORFs in generated sequences using Prodigal v2.6.3 with default parameters in metagenome\nmode (-p meta) (Hyatt et al., 2010), generated proteins were aligned against the full-length prompt pro-\ntein sequence using MAFFT (v7.526) (Katoh et al., 2002) for sequence identity calculations. For evaluating\nsequence degeneracy, DustMasker (version 2.14.1+galaxy2) (Morgulis et al., 2006;Camacho et al., 2009;\nCock et al., 2015) was run across the full-length generations using default parameters and the proportion of\nmasked nucleotides was calculated.',
    'Final parameters (temperature = 0.7, top-k = 4, top-p= 1.0) were se-\nlected based on maximizing sequence completion accuracy while maintaining DustMasker proportions below\n0.2, a value chosen to be just slightly higher than the typical frequency of non-coding DNA in prokaryotic\ngenomes (Rogozin et al., 2002;Troyanskaya et al., 2002). Sequence completion prompt compilation and evaluation\nSequences of highly-conserved genes from across prokaryotic biology were downloaded in FASTA format from\nNCBI GenBank (Clark et al., 2015). Selected genes included rpoSfrom E. coliK-12 (GenBank: NC_000913.3,\ncoordinates2867551-2868753), gyrAfrom S. typhimurium LT2(GenBank: NC_003197.2,coordinates2337442-\n2339871),and ftsZ1from H. volcanii DS2(GenBank: NC_013967.1,coordinates1675932-1677149).',
    'Prompts\nwere prepared by extracting 30%, 50%, and 80% sequence lengths from the 5‚Äô end. Sequence completion performance was evaluated across varying prompt lengths (30%, 50%, and 80%\nof input sequence) using optimal sampling parameters for the Evo 1.5, Evo 1 8k (previous version of Evo\ntrained with context length of 8,192 tokens), and Evo 1 131k (previous version of Evo trained with context\nlength of 131,072 tokens) models (temperature = 0.7, top-k = 4, top-p = 1) (Nguyen et al., 2024). For each\nprompt, one hundred sequences of length 2,500 nt were generated.',
    'Prompts were subsequently appended\nto the start of each generated sequence and ORFs identified using Prodigal v2.6.3 with default parameters in\nmetagenome mode (-p meta). Generated proteins were then aligned against their corresponding wild-type\nsequences using MAFFT (v7.526) using default parameters for sequence identity calculations. Operon completion prompt compilation and evaluation\nSequences encoding the trpoperon and modABC operon from E. coliK-12 (GenBank: NC_000913.3) were\ndownloaded in FASTA format from NCBI GenBank.',
    'For modABC, prompts were prepared from the full coding\nsequences for modA(coordinates 795089-795862), modB(coordinates 795862-796551), modC(coordinates\n796554-797612) and acrZ(coordinates 794773-794922) (Clark et al., 2015). For trp, prompts were pre-\npared from the full coding sequences for trpE(coordinates c1322946-1321384), trpD(coordinates c1321384-\n14. CC-BY 4.0 International license perpetuity.',
    'It is made available under apreprint (which was not certified by peer review) is the author/funder, who has granted bioRxiv a license to display the preprint in The copyright holder for this this version posted December 18, 2024. ; https://doi.org/10.1101/2024.12.17.628962doi: bioRxiv preprint 1319789), trpC(coordinates c1319788-1318427), trpB(coordinates c1318415-1317222), and trpA(coor-\ndinates c1317222-1316416). For bidirectional generation testing, the reverse complement of each prompt\nsequence was generated using Biopython (Cock et al., 2009). Sequence completion performance was evaluated across the compiled operon completion prompts using\npreviously identified optimal sampling parameters for Evo 1.5.',
    'For each prompt, 1000 sequences of length\n2500 nt were generated. Following identification of ORFs using Prodigal, directional completion was assessed\nby searching trpE-prompted generations for trpD-like ORFs, trpDreverse complement-prompted generations\nfortrpE-like ORFs, and similar pairing combinations across both modABC andtrpoperons. Protein sequences\nwere then aligned against their corresponding wild-type proteins for sequence identity calculations using\nMAFFT (v7.526).',
    'Structural similarity was evaluated by generating protein structure predictions using Al-\nphaFold 3 (Abramson et al., 2024) for both generated and wild-type sequences, with structural alignments\nand TM-score calculations performed using TM-align (Zhang and Skolnick, 2005). Native and predicted pro-\ntein structures were subsequently visualized using ChimeraX (Pettersen et al., 2021). Positional entropy evaluation\nPer-position amino acid and nucleotide entropies were calculated from multiple sequence alignments of 500\ngenerated and native modBsequences.',
    'Native modBsequences were fetched by querying ‚ÄòmodB‚Äô in NCBI\nprotein, filtering by bacteria, and downloading the corresponding amino acid and nucleotide sequences in\nformat ‚ÄòFASTA‚Äô and ‚ÄòFASTA CDS‚Äô respectively. Generated modBsequences were taken by selecting a random\nsample of 500 modB ORFs from the modBsequences generated by prompting with modAduring the operon\ncompletion evaluation. First, nucleotide and amino acid sequences were aligned with MAFFT (v7.526) and\ntrimmed to remove gaps appearing in >80% of sequences.',
    'For each position ùëñ, the Shannon entropy was\nthen calculated as ùêª(ùë•ùëñ)=‚àí√ç\nùë•ùëñ‚ààXùëù(ùë•ùëñ)log2(ùëù(ùë•ùëñ))and normalized by dividing the calculated entropy by the\nmaximum Shannon entropy (2 for nucleotides meaning all 4 bases are equally present, 4.32 for amino acids\nmeaningall20standardaminoacidsareequallypresent), where ùëù(ùë•ùëñ)representsthefrequencyofeachamino\nacid or nucleotide ùë•ùëñat that position and Xindicates the vocabulary. Toxin-antitoxin prompt compilation\nGenomiclociandsequencesencodingTypeIItoxin-antitoxinsystemsequenceswereobtainedbydownloading\nthe nucleotide sequence information for all experimentally validated type II-TAs from the TADB 3.0 database\n(Guan et al., 2024). Using NCBI‚Äôs Entrez Programming Utilities API (EFetch from the ‚Äônuccore‚Äô database using\nthe genomic loci from TADB 3.0) (Sayers et al., 2022), the 500 bp of upstream and downstream flanking\nsequence were extracted for each T2TA locus.',
    'In total, for each T2TA system, six types of prompts were\nprepared: (1) individual toxin sequences, (2) individual antitoxin sequences, (3) the upstream context of\nthe toxin loci, (4) the downstream context of the toxin loci, (5) the upstream context of the antitoxin loci,\nand (6) the downstream context of the antitoxin loci. Following successful identification of an Evo generated\ntoxin (see section ‚ÄúEvaluation of toxin activity ‚Äù below), conjugate antitoxins were subsequently generated\nvia prompting with the generated DNA sequence encoding the toxin. Toxin-antitoxin sampling and filtering\nTo generate T2TA candidates, we first generated 57,936 sequences of 2,000 nucleotides each using Evo 1.5\n(temperature = 0.7, top-k = 4, top-p = 1.0) with our compiled T2TA prompts.',
    'A multi-stage filtering pipeline\nwas then applied to identify promising candidates. First, Prodigal was used to identify ORFs, excluding se-\nquences containing proteins over 300 amino acids or those with only a single ORF. Next, SegMasker (version\n2.14.1+galaxy2) with default parameters (Camacho et al., 2009;Cock et al., 2015) was employed to remove\nsequences containing low complexity regions with limited amino acid diversity.',
    'Wethenassessedtheprotein-proteininteractionpotentialofco-generatedORFsbyco-foldingallORFpairs\nwithineachremaininggenerationusingESMFold. Generationswereretainediftheycontainedpairedproteins\nwith pDockQ (Basu and Wallner, 2016;Bryant et al., 2022) scores above 0.23 and high sequence dissimilar-\nity between components. To identify novel candidates, the remaining sequences were searched against the\nnon-redundant protein sequence database using BLAST (E-value cutoff 5√ó10‚àí2), selecting for generations\ncontaining at least one component with no significant BLAST hits to known toxins or antitoxins.',
    'Final candi-\n15. CC-BY 4.0 International license perpetuity. It is made available under apreprint (which was not certified by peer review) is the author/funder, who has granted bioRxiv a license to display the preprint in The copyright holder for this this version posted December 18, 2024. ; https://doi.org/10.1101/2024.12.17.628962doi: bioRxiv preprint dates were then selected based on high confidence interaction prediction using AlphaFold 2 (Jumper et al.,\n2021).',
    'Following the identification of functional toxin candidates via experimental testing, the Evo-generated\nsequence encoding the strongest Evo-generated toxin, EvoRelE1, was used as a prompt to generate further\ndiversified antitoxin candidates. After generating a total of 3,000 sequences from the EvoRelE1 prompt, gen-\nerated ORFs were cleaned and filtered as above before being co-folded with EvoRelE1. As with the first round\nof generations, candidates were filtered for high pDockQ scores and low-sequence homology to known anti-\ntoxins before being evaluated for strong predicted co-folds using AlphaFold2.',
    'Final antitoxin candidates were\nfurther characterized using Foldseek web server (van Kempen et al., 2024) searches of the AlphaFold3 pre-\ndicted structures (probability threshold of 0.6), blastp searches against the non-redundant protein database\n(E value threshold of 1), and HHpred searches (probability threshold of >90%) (S√∂ding et al., 2005). Cloning of toxin and antitoxin sequences\nSequences encoding Evo-generated toxins and antitoxins were codon optimized for E. coliexpression using\nIDT‚Äôs codon optimization tool and synthesized as eBlocks. The toxin plasmid backbone was prepared by PCR\namplification (New England Biosciences) and gel purification (Qiagen) of pAraSpCas9+spMu (Forsberg et al.,\n2019) to exclude the Cas9, crRNA, tracrRNA, and gRNA sequences (Data S1), creating an empty arabinose-\ninducible vector with spectinomycin resistance.',
    'Codon-optimized toxin sequences were subsequently inserted\nintothemodifiedbackboneusingGibsonAssembly(NewEnglandBiosciences)accordingtothemanufacturer‚Äôs\nrecommendations. Forantitoxincloning, thepZE21_tetR-AcrIIA4_kanRvector(Forsbergetal., 2019)(DataS1 )wasdigested\nwith EcoRI and HindIII (New England Biosciences) and gel purified (Qiagen) to remove the AcrIIA4 sequence. Codon-optimizedantitoxinsequencesweretheninsertedintothedigestedvectorusingGibsonAssembly(New\nEngland Biosciences).',
    'Assembled plasmids were transformed into chemically competent Stellar E. coliHST08 cells (Takara Bio-\nsciences), and positive clones were selected on LB agar plates containing either 50 g/mL spectinomycin (for\nthe toxin constructs) or 50 g/mL kanamycin (for the antitoxin constructs). Plasmid sequences were then\nconfirmed by Sanger sequencing (Elim Biosciences). Evaluation of toxin activity\nToxinactivitywasassessedthroughgrowthinhibitionassaysin E. coliNEBTurbocells.',
    'Cellstransformedwith\ntoxin constructs were first grown overnight in LB medium containing 50 g/mL spectinomycin. Cultures were\nthen diluted 1:10 into 1 mL fresh LB supplemented with 50 g/mL spectinomycin and 2 mg/mL arabinose to\ninduce toxin expression. After 2 hours of induction, cultures were further diluted 1:40 into 200 L of the same\nmedium in triplicate wells of a 96-well plate.',
    'Growth was monitored using a Tecan Spark plate reader at 37¬∞C\nwith orbital shaking, measuring optical density (OD 600) at 30-minute intervals over an 8-hour period. Growth\ncurves were analyzed to evaluate the extent of toxin-mediated growth inhibition. Evaluation of antitoxin activity\nFor antitoxin evaluation, E. coliNEB Turbo cells were co-transformed with both toxin and antitoxin constructs\nand selected on LB plates containing 50 g/mL spectinomycin and 50 g/mL kanamycin.',
    'Co-transformed cells\nwere grown overnight in LB medium containing both antibiotics. Cultures were then diluted 1:10 into 1 mL\nfresh LB supplemented with both antibiotics and 2 mg/mL arabinose to induce toxin expression. After 2 hours\nof induction, cultures were further diluted 1:40 into 200 L of the same medium in triplicate wells of a 96-\nwell plate.',
    'Growth was monitored using a Tecan Spark plate reader at 37¬∞C with orbital shaking, measuring\nopticaldensity(OD 600)at30-minuteintervalsoveran8-hourperiod. Growthcurveswereanalyzedtoevaluate\nantitoxin-mediated rescue of growth. Statistical comparisons were done using a one-sided Student‚Äôs t-test against the EvoRelE1+ eGFP negative\ncontrol, with ùëù-values for EvoAT1-4 being 1.61√ó10‚àí7,1.42√ó10‚àí7,2.30√ó10‚àí5, and 1.69√ó10‚àí5respectively.',
    '16. CC-BY 4.0 International license perpetuity. It is made available under apreprint (which was not certified by peer review) is the author/funder, who has granted bioRxiv a license to display the preprint in The copyright holder for this this version posted December 18, 2024. ; https://doi.org/10.1101/2024.12.17.628962doi: bioRxiv preprint Anti-CRISPR prompt compilation\nGenomic sequences containing known Type II Cas9-targeting anti-CRISPR (acr ) genes and their associated\noperons were obtained from previously characterized anti-CRISPR systems annotated in AcrDB.',
    'Using the\nEntrez API, we extracted Acr coding sequences along with 500 bp of flanking sequence both upstream and\ndownstream of each Acr locus. For each Acr system, we prepared six types of prompts: (1) individual acr\nsequences, (2) anti-CRISPR associated (aca) gene sequences, (3) the upstream context of the Acr loci, (4) the\ndownstream context of the Acr loci, (5) the upstream context of the acaloci, and (6) the downstream context\nof the acaloci. Anti-CRISPR sampling and filtering\nTogenerateAcrcandidates,wesampledatotalof3,160sequencesof1,000nucleotideseachusingEvo1.5with\nour compiled Acr prompts.',
    'A multi-stage filtering pipeline was then applied to identify promising candidates. First,ProdigalwasusedtoidentifyORFs,excludingsequencescontainingproteinsover200aminoacids. Next,\nSegMasker was employed to remove sequences containing over 50% low complexity regions or limited amino\nacid diversity.',
    'Sequences were subsequently folded using ESMFold to remove any candidates with very low\nconfidence folds (pLDDT < 0.25) or no secondary structure. Candidate sequences were then evaluated using\nPaCRISPR, a machine learning model trained to identify potential anti-CRISPR proteins based on sequence\nfeatures(Wangetal., 2020). FollowingguidelinesestablishedbyPaCRISPR,candidatesscoringabove0.5were\nconsidered potential anti-CRISPRs, with sequences scoring above 0.75 advancing to the next step.',
    'Generated\nsequences classified as potential Acrs by PaCRISPR were then searched against the non-redundant protein\nsequence database using blastp (E-value cutoff of 1) to identify candidates with varying degrees of sequence\nnovelty. For comparison, we also applied this filtration pipeline to evaluate sequences generated by prompting\nwith randomly generated DNA sequences and non-Acr related genomic sequences. We then assessed the\nsequence diversity among the predicted Acrs by performing pairwise alignments using MAFFT (v7.526) on a\nset of 56 randomly selected sequences that scored above 0.75 in PaCRISPR.',
    'The resulting pairwise sequence\nidentities were visualized using a matplotlib heatmap. Cloning of anti-CRISPR sequences\nSequences encoding Evo-generated anti-CRISPR proteins were codon optimized for E. coliexpression using\nIDT‚Äôs codon optimizationtool and synthesized aseBlocks. Togenerate the cloning backbone, the pZE21_tetR-\nAcrIIA4-Coli_kanRvector(Forsbergetal., 2019)(DataS1)wasdigestedwithEcoRIandHindIII(NewEngland\nBiosciences) to remove the AcrIIA4 sequence and gel purified (Qiagen).',
    'Codon-optimized Acr sequences were\nthen inserted into the digested vector using Gibson Assembly (New England Biosciences) according to the\nmanufacturer‚Äôs recommendations. Assembled plasmids were then transformed into chemically competent\nStellar E. coliHST08 cells (Takara Biosciences), and positive clones were selected on LB agar plates containing\n50 g/mL kanamycin. Plasmid sequences were confirmed by Sanger sequencing (Elim Biosciences).',
    'Liquid culture assay for measuring anti-CRISPR activity\nAnti-CRISPR activity was assessed through protection assays against SpCas9-mediated DNA cleavage in E.\ncoli. NEBTurbocellswerefirstco-transformedwithboththeAcrexpressionplasmidandtheCRISPRtargeting\nplasmidcontainingSpCas9andaKanRtargetingguideRNA(Forsbergetal., 2019)(pAraCas9+Sp2+Sp6+\nI-SceI,Data S1) and grown overnight in LB medium containing both 50 g/mL kanamycin and spectinomycin. For liquid culture survival assays, 30 L of overnight culture was diluted into 1000 L fresh LB medium\ncontaining spectinomycin and 0.2 mg/mL arabinose to induce SpCas9 expression and deplete the kanamycin\nresistance plasmid.',
    'After 7 hours of growth, cultures were normalized to equal optical density and diluted\n1:40 into 200 L fresh LB medium containing 50 g/mL kanamycin and spectinomycin in a 96-well plate to\nselect for cells with active anti-CRISPR proteins. Growth was monitored using a Tecan Spark plate reader at\n37¬∞C with orbital shaking, measuring optical density (OD 600) at 30-minute intervals over a 7-hour period. For\neach Acr, an uninduced control without arabinose was used to normalize growth values.',
    'Statistical comparisons were done using a one-sided Student‚Äôs ùë°-test against the random sequence negative\ncontrol and AcrIIA2, with ùëù-values for EvoAcr1-5 being 8.31√ó10‚àí6,6.19√ó10‚àí6,8.19√ó10‚àí6,1.43√ó10‚àí6,\n17. CC-BY 4.0 International license perpetuity. It is made available under apreprint (which was not certified by peer review) is the author/funder, who has granted bioRxiv a license to display the preprint in The copyright holder for this this version posted December 18, 2024. ; https://doi.org/10.1101/2024.12.17.628962doi: bioRxiv preprint and8.18√ó10‚àí7against the negative control respectively and the ùëù-value for EvoAcr5 against AcrIIA2 being\n5.07√ó10‚àí4.',
    'Phage plaque assay for measuring anti-CRISPR activity\nForphageprotectionassays, E. coliNEBTurbocellswereco-transformedwithboththeAcrexpressionplasmid\nandamodifiedSpCas9plasmidcontainingaguideRNAtargetingT4(GT7)phage(SpCas9-mrh2, DataS1)ata\n5:1 Acr:Cas9 ratio. Co-transformed cells were grown overnight in LB medium containing 50 g/mL kanamycin\nand spectinomycin. Cultures were then diluted 1:10 into fresh LB medium supplemented with spectinomycin,\nkanamycin, and0.2mg/mLarabinosetoinduceCas9expression.',
    'WhenculturesreachedanOD 600of0.4, 300\nL was mixed into 10 mL of 0.7% soft agar containing 50 mg/mL kanamycin, 50 mg/mL spectinomycin, and\n0.02 mg/mL arabinose. Plates were allowed to harden and T4(GT7) phage at a titer of 1.7√ó108PFU/mL was\nserially diluted 1:5 and spotted onto the bacterial lawn. Plates were then incubated overnight at 37¬∞C before\nbeing imaged to visualize plaque formation.',
    'Measurement of Evo-generated protein sequence diversity\nSequence and structural diversity of generated toxin-antitoxin pairs and anti-CRISPRs were assessed through\na combination of sequence and structure-based searches. Protein sequences were searched against the NCBI\nnon-redundant protein database using blastp (E-value threshold of 1.0) to identify potential homologs. An\nadditionalsearchagainstOpenGenomewasperformedusingMMseqs2(version15.6f452)withmaximumsen-\nsitivity(-s7)andotherparameterssettothedefaultvalues(SteineggerandS√∂ding, 2017)toevaluatesimilar-\nity to sequences in the training data.',
    'Structural similarity was evaluated by searching AlphaFold 3-predicted\nstructures against the BFVD, AFDB-Proteome, AFDB-Swissprot, AFDB-50, BFMD, CATH50, GMGCL_ID, MG-\nNIFY_ESM30, andPDB100proteindatabasesusingtheFoldseekwebserver (https://search.foldseek. com/search) (vanKempenetal., 2024). Foreachgeneratedprotein,closestsequenceandstructuralmatches\nwere identified from all three searches (BLAST, MMseqs2, and Foldseek) and evaluated to determine the\ndegree of novelty compared to known proteins.',
    'Pairwise sequence identities were calculated from BLAST\nalignments, while structural similarities were quantified using Foldseek TM-scores. SynGenome prompt compilation\nProtein sequences from prokaryotic and phage organisms were retrieved from UniProt, with all Swissprot pro-\nteins with associated coding regions and a random sample of TrEMBL proteins with associated coding regions\nbeing used as starting points for prompts (Consortium, 2024). Genomic contexts were retrieved via NCBI‚Äôs\nEntrez Programming Utilities (E-utilities) API, specifically using EFetch with the ‚Äônuccore‚Äô database using the\nCDS annotations associated with the Protein Sequence accession in UniProt.',
    'For each protein‚Äôs associated\ngenomic identifier, we constructed three API calls: one to extract the coding sequence using sequence feature\ncoordinates, and two to extract the flanking regions by calculating positions 500 nucleotides upstream and\ndownstream of the CDS boundaries. For CDS regions that were longer than 500 nucleotides, the prompt was\nderived from the final 500 nucleotides in the coding sequence. For CDS regions that were shorter than 500\nnucleotides, the CDS sequence was trimmed to the nearest 100 bp.',
    'Additionally, we generated reverse com-\nplement sequences for each region using the Biopython Seq module, resulting in six distinct prompts per pro-\ntein: upstream, coding sequence (CDS), downstream, and their respective reverse complements. Associated\nfunctional annotations including Gene Ontology (GO) terms (Ashburner et al., 2000;The Gene Ontology Con-\nsortium et al., 2023), species names, and InterPro domains (Blum et al., 2024) were retrieved from UniProt‚Äôs\nREST API for each protein using their UniProt accession numbers and linked to their corresponding prompts. SynGenome sampling\nSequences were generated using Evo 1.5 with optimized sampling parameters (temperature = 0.7, top-k = 4,\ntop-p= 1.0).',
    'For each prompt type, we generated 2 sequences of 5,000 nucleotides in length, yielding a total\ndatabase size of over 120 billion base pairs. Generation was performed in parallel across multiple compute\nnodes to facilitate large-scale sequence production. 18.',
    'CC-BY 4.0 International license perpetuity. It is made available under apreprint (which was not certified by peer review) is the author/funder, who has granted bioRxiv a license to display the preprint in The copyright holder for this this version posted December 18, 2024. ; https://doi.org/10.1101/2024.12.17.628962doi: bioRxiv preprint Data filtering of SynGenome sequences\nGenerated sequences underwent a multi-step filtering pipeline to remove low-complexity regions while pre-\nserving biologically plausible features. Initial validation removed any invalid characters from the nucleotide\nsequences.',
    'These sequences were then processed using DustMasker (NCBI BLAST+ v2.16.0) with a masking\nlevel of 30 (-level 30) and FASTA output format (-outfmt fasta) to identify low-complexity regions. Following\nDustMasker processing, we implemented two additional filtering steps: removal of successive 100 nucleotide\nchunks from the sequence end if they contained more than 40% masked bases, and elimination of any contin-\nuous masked regions longer than 800 nucleotides. Sequences were excluded from the final database based on\nseveral criteria: length below 100 nucleotides, masked base content exceeding 80% in sequences shorter than\n2,000 nucleotides, complete masking of all bases, or empty/NaN values.',
    'The sequences passing these filtering\ncriteria were converted to uppercase and retained in the final database. Prediction of SynGenome ORFs\nOpen reading frames were identified in the filtered sequences using Prodigal v2.6.3 with default parame-\nters and metagenomic prediction mode (-p meta). Predictions were initially refined by excluding sequences\nshorter than 40 amino acids or longer than 1200 amino acids and sequences with incomplete protein se-\nquences.',
    'Following basic filtration, low-complexity regions were identified using NCBI SegMasker (window\nsize 15, locut 1.8, hicut 3.4), with sequences containing >20% masked regions being excluded. Additional\ncomplexity filters removed sequences with fewer than 12 unique amino acids or highly repetitive k-mer pat-\nterns (k=3-10, threshold >40% coverage). This multi-step filtering process ensured the retention of high-\nquality protein predictions while removing potentially spurious or low-complexity sequences.',
    'Creation of SynGenome website\nThe SynGenome website was implemented with HTML, CSS, and JavaScript to provide a searchable web\ninterface at https://evodesign.org/syngenome/. The interface allows users to query sequences using\nprotein names, UniProt IDs, InterPro domains, species names, or GO terms. The database was structured to\nmaintain associations between generated sequences and their corresponding prompt annotations.',
    'The raw\nSynGenome data is hosted on Hugging Face for public access. Creation of SynGenome UMAP and Leiden Clusters\nTo generate the UMAP, first, a random sample of 50,000 sequences encoding at least one ORF with prompts\nderived from the CDS sequence was extracted from SynGenome. This random sample was subsequently fil-\ntered to remove generations with < 40% or > 60% GC content, resulting in a final set of 36,762 generations.',
    'Embeddings were generated for both the prompt and corresponding generated sequences by extracting ac-\ntivations from the MLP layer 3 in the 6th hyena block of Evo 1.5‚Äôs architecture before being mean pooled\nalong the sequence dimension to create fixed-length representations. These high-dimensional embeddings\nwere subsequently reduced to two dimensions using UMAP (n_neighbors=15, min_dist=0.1). Sequences\nwere colored according to their percent identity to sequences in the training data, calculated using MMseqs2\nagainst a database of prokaryotic genomes used in model training.',
    'These high-dimensional embeddings were\nreduced to two dimensions using UMAP with ScanPy (version 1.10.3) (Wolf et al., 2018) default parameters. Sequences were colored according to their percent identity to sequences in the training data, calculated using\nMMseqs2 against OpenGenome. Graph-based clustering was performed using the default Leiden algorithm\nimplemented in Scanpy (version 1.10.3).',
    'The resolution parameter was optimized by evaluating cluster sta-\nbility across resolutions from 0.1 to 0.5, measuring both the number of clusters and coefficient of variation of\ncluster sizes. A final resolution of 0.2 was selected for clustering based on these metrics. Comparison of SynGenome to native prokaryotic sequences\nTo evaluate how representative SynGenome was of natural prokaryotic sequences, first, a random sample of\n36,762 5,000-nt generations was taken from OpenGenome to match the number of sequences used in the rep-\nresentative sample of SynGenome.']


@app.route('/start', methods=['POST'])
def start_task():
    data = request.json
    content = data.get('content')
    schedule_type = data.get('schedule_type')  # 'time' Êàñ 'interval'
    time = data.get('time')  # ÂÆöÊó∂Êó∂Èó¥ÔºåÊ†ºÂºè‰∏∫ISOÂ≠óÁ¨¶‰∏≤
    interval = data.get('interval')  # Èó¥ÈöîÊó∂Èó¥ÔºåÂçï‰Ωç‰∏∫Áßí

    if schedule_type == 'time':
        run_date = parse(time)
        job_id = str(uuid.uuid4())
        scheduler.add_job(send_tweet, 'date', run_date=run_date, args=[content], id=job_id)
    elif schedule_type == 'interval':
        job_id = str(uuid.uuid.uuid4())
        scheduler.add_job(send_tweet, 'interval', seconds=interval, args=[content], id=job_id)
    else:
        return jsonify({'message': 'Êó†ÊïàÁöÑÂÆöÊó∂Á±ªÂûã'}), 400

    return jsonify({'message': '‰ªªÂä°Â∑≤ÂêØÂä®', 'job_id': job_id})


@app.route('/stop', methods=['POST'])
def stop_task():
    data = request.json
    job_id = data.get('job_id')
    if job_id:
        try:
            scheduler.remove_job(job_id)
            return jsonify({'message': '‰ªªÂä°ÂèñÊ∂àÊàêÂäü'})
        except Exception as e:
            return jsonify({'message': f'ÂèñÊ∂à‰ªªÂä°Â§±Ë¥•: {str(e)}'}), 500
    else:
        return jsonify({'message': 'Êó†ÊïàÁöÑ job_id'}), 400


@app.route('/tasks', methods=['GET'])
def get_tasks():
    jobs = scheduler.get_jobs()
    tasks = []
    for job in jobs:
        tasks.append({
            'job_id': job.id,
            'next_run_time': job.next_run_time.isoformat() if job.next_run_time else None,
            'content': job.args[0] if job.args else 'Êó†ÂÜÖÂÆπ',
        })
    return jsonify({'tasks': tasks})


@app.route('/send', methods=['POST'])
def send_now():
    data = request.json
    content = data.get('content')
    send_tweet(content)
    return jsonify({'message': 'Êé®ÊñáÂ∑≤ÂèëÈÄÅ'})


def send_tweet(content):
    try:
        print(twitter_client.create_tweet(text=content))
        print(f"ÊàêÂäüÂèëÂ∏ÉÊé®ÊñáÔºö{content}")
    except Exception as e:
        print(f"ÂèëÂ∏ÉÊé®ÊñáÊó∂Âá∫ÈîôÔºö{str(e)}")


@app.route('/random_text', methods=['GET'])
def random_text():
    if paragraphs:
        random_paragraph = random.choice(paragraphs)
        random_paragraph = translate_text(random_paragraph)
        return jsonify({'text': random_paragraph})
    else:
        return jsonify({'message': 'Ê≤°ÊúâÂèØÁî®ÁöÑÊñáÊú¨ÊÆµËêΩ'}), 404


@app.route('/')
def index():
    return 'Á®ãÂ∫èÊ≠£Âú®ËøêË°å\nËØ∑ËÆøÈóÆÂâçÁ´ØÈ°µÈù¢'


if __name__ == '__main__':
    app.run(debug=True)
